packages/test/test-utils/src/testContainerRuntimeFactoryWithDefaultDataStore.ts

```diff
@@ -23,35 +24,72 @@ const getDefaultFluidObject = async (runtime: IContainerRuntime) => {
 };
 
 /**
- * ! Note: This function is purely needed for back-compat as the constructor argument structure was changed
+ * {@link ContainerRuntimeFactoryWithDefaultDataStoreConstructor} input properties.
+ *
+ * @internal
+ */
+export interface ContainerRuntimeFactoryWithDefaultDataStoreProps {
+	defaultFactory: IFluidDataStoreFactory;
+	/**
+	 * The data store registry for containers produced.
+	 */
+	readonly registryEntries: NamedFluidDataStoreRegistryEntries;
+
+	readonly dependencyContainer?: IFluidDependencySynthesizer;
+
+	/**
+	 * Request handlers for containers produced.
+	 * @deprecated Will be removed once Loader LTS version is "2.0.0-internal.7.0.0". Migrate all usage of IFluidRouter to the "entryPoint" pattern. Refer to Removing-IFluidRouter.md
+	 */
+	// eslint-disable-next-line import/no-deprecated
+	readonly requestHandlers?: RuntimeRequestHandler[];
+
+	/**
+	 * The runtime options passed to the IContainerRuntime when instantiating it
+	 */
+	readonly runtimeOptions?: IContainerRuntimeOptions;
+
+	/**
+	 * Function that will initialize the entryPoint of the IContainerRuntime instances
+	 * created with this factory
+	 */
+	readonly provideEntryPoint?: (runtime: IContainerRuntime) => Promise<FluidObject>;
+}
+
+/**
+ * {@link @fluidframework/container-definitions#IRuntimeFactory} construct signature.
+ *
+ * @internal
+ */
+export type ContainerRuntimeFactoryWithDefaultDataStoreConstructor = new (
+	props: ContainerRuntimeFactoryWithDefaultDataStoreProps,
+) => IRuntimeFactory;
+
+/**
+ * @remarks
+ * This function is purely needed for back-compat as the constructor argument structure of
+ * `ContainerRuntimeFactoryWithDefaultDataStore` was changed.
+ *
  * @internal
  */
 export const createContainerRuntimeFactoryWithDefaultDataStore = (
-	Base: typeof ContainerRuntimeFactoryWithDefaultDataStore = ContainerRuntimeFactoryWithDefaultDataStore,
-	ctorArgs: {
-		defaultFactory: IFluidDataStoreFactory;
-		registryEntries: NamedFluidDataStoreRegistryEntries;
-		dependencyContainer?: any;
-		// eslint-disable-next-line import/no-deprecated
-		requestHandlers?: RuntimeRequestHandler[];
-		runtimeOptions?: IContainerRuntimeOptions;
-		provideEntryPoint?: (runtime: IContainerRuntime) => Promise<FluidObject>;
-	},
-): ContainerRuntimeFactoryWithDefaultDataStore => {
+	ctor: ContainerRuntimeFactoryWithDefaultDataStoreConstructor,
+	ctorProps: ContainerRuntimeFactoryWithDefaultDataStoreProps,
+): IRuntimeFactory => {
 	try {
-		return new Base(ctorArgs);
+		return new ctor(ctorProps);
 	} catch (err) {
-		// IMPORTANT: The constructor argument structure changed, so this is needed for dynamically using older ContainerRuntimeFactoryWithDefaultDataStore's
+		// IMPORTANT: The constructor argument structure changed, so this is needed for dynamically using older `ContainerRuntimeFactoryWithDefaultDataStore`s
```

Comment:
this seems super questionable. A failing assert in the constructor would cause is to suppress the error and run it again with different arg setups?

DO we have to keep this?
I don't feel qualified to review touching anything related to this workaround without a lot more knowledge as this seems super likely to break in ways I wouldn't notice during review.

Reply:
> this seems super questionable. A failing assert in the constructor would cause is to suppress the error and run it again with different arg setups?

Agreed.

> DO we have to keep this?

No idea. But note that this behavior is unchanged in this PR. Messing with these behaviors is beyond the scope of what I'm trying to do here. If someone can offer insight into whether or not this is necessary, I am happy to clean up. But otherwise, I'm going to leave this alone.

Reply:
Per Kian's comment [here](https://github.com/microsoft/FluidFramework/pull/23848#discussion_r1972619481), this logic is still needed. But as previously noted, cleaning this up is beyond the scope of this PR.